File: src/mcu/rtc.rs
//==============================================================================
// Notes
//==============================================================================
// mcu::rtc.rs

//==============================================================================
// Crates and Mods
//==============================================================================
use nrf52832_pac;
use nrf52832_pac::interrupt;
use crate::drivers::debug;

//==============================================================================
// Enums, Structs, and Types
//==============================================================================
#[allow(dead_code)]
pub enum WakeInterval {
	Interval125MS	= 512,
	Interval250MS	= 1024,
	Interval500MS	= 2048,
	Interval1S 		= 4096
}

//==============================================================================
// Macros
//==============================================================================


//==============================================================================
// Variables
//==============================================================================
static mut _INITIALIZED: bool = false;
static mut _WAKE_INTERVAL: u32 = 0;
static mut _FRACTION: u32 = 0;
static mut _SECONDS: u32 = 0;

//==============================================================================
// Implementations
//==============================================================================
#[allow(dead_code)]
pub fn init(p: &nrf52832_pac::Peripherals, interval: WakeInterval) {
	if unsafe { _INITIALIZED } {
		return;
	}

	unsafe { _WAKE_INTERVAL = interval as u32; }

	enable(p, true);

	unsafe { _INITIALIZED = true; }
}

#[allow(dead_code)]
pub fn get_timestamp() -> u32 {
	unsafe { _SECONDS }
}

#[allow(dead_code)]
pub fn get_timestamp_fraction() -> u32 {
	unsafe { _FRACTION }
}

fn enable(p: &nrf52832_pac::Peripherals, is_enabled: bool) {
	let clock = &p.CLOCK;
	let rtc = &p.RTC0;

	nrf52832_pac::NVIC::mask(nrf52832_pac::Interrupt::RTC0);

	if is_enabled {
		// Enable XTAL for Low Freq Clock Source
		clock.lfclksrc.write(|w| w.src().xtal());
		clock.tasks_lfclkstart.write(|w| unsafe { w.bits(1) });
		//TODO: waiting indefinitely here
		while clock.events_lfclkstarted.read().bits() == 0 {};

		//Disable RTC
		rtc.tasks_stop.write(|w| unsafe { w.bits(1) });
		
		//prescale by 8 : 32768 / 8 = 4096 Hz
		rtc.prescaler.write(|w| unsafe { w.prescaler().bits(7) });
	

		//define the wake interval
		rtc.cc[0].write(|w| unsafe { w.bits(_WAKE_INTERVAL) });

		//connect the interrupt event signal on compare0 match
		rtc.intenset.write(|w| w.compare0().set_bit());

		unsafe {
			nrf52832_pac::NVIC::unpend(nrf52832_pac::Interrupt::RTC0);
			nrf52832_pac::NVIC::unmask(nrf52832_pac::Interrupt::RTC0);
		}

		//Enable RTC
		rtc.tasks_start.write(|w| unsafe { w.bits(1) });
	}

	unsafe { 
		_SECONDS = 0;
		_FRACTION = 0;
	}
}

// =============================================================================
// Interrupt Handler
//==============================================================================
#[interrupt]
fn RTC0() {
	unsafe { 
		let rtc = nrf52832_pac::Peripherals::steal().RTC0;
	
		if rtc.events_compare[0].read().bits() > 0 {
			_FRACTION += _WAKE_INTERVAL;
		
			if _FRACTION >= WakeInterval::Interval1S as u32 {
				_SECONDS += _FRACTION / WakeInterval::Interval1S as u32;
				_FRACTION = 0;
			}
		}

		rtc.events_compare[0].write(|w| w.bits(0));
		rtc.tasks_clear.write(|w| w.bits(0));
	}
}

//==============================================================================
// Task Handler
//==============================================================================
pub fn task_handler(){
	static mut LAST_SECONDS: u32 = 0;
	
	unsafe {
		if LAST_SECONDS != get_timestamp() {
			LAST_SECONDS = get_timestamp();
			// let time = "Time: ".as_bytes() + debug::number_to_string(&last_seconds);
			debug::push_log("Next second: ");
		}
	}
}

File: nrf52832_lib/src/rtc.c
//==============================================================================
// Notes
//==============================================================================
// rtc.c

//==============================================================================
// Includes
//==============================================================================
#include "rtc.h"

//==============================================================================
// Defines
//==============================================================================


//==============================================================================
// Types
//==============================================================================
typedef void (*scheduleCbFunction)(RtcSchedule_t *schedule);

//==============================================================================
// Private Prototypes
//==============================================================================
static void rtc_initMcu(uint8_t enable);
static inline uint16_t rtc_getFracTimeMcu();
static void rtc_scheduleAdd(RtcSchedule_t *schedule);
static uint8_t rtc_scheduleExists(RtcSchedule_t *schedule);
static uint32_t rtc_scheduleGetNextTime(uint32_t interval);

//==============================================================================
// Variables
//==============================================================================
static uint8_t _initialized = false;
static volatile uint32_t _seconds = 0;
static volatile uint32_t _fraction = 0;
static RtcWakeInterval_t _interval;
static RtcSchedule_t *_firstSchedule = NULL;

//==============================================================================
// Public Functions
//==============================================================================
void RTC_Init(RtcWakeInterval_t interval){
	_interval = interval;

	rtc_initMcu(true);
	
	_initialized = true;
}

uint32_t RTC_GetCurrentTime(){
	return _seconds;
}

uint16_t RTC_GetCurrentFractionalMs(){
	float frac = rtc_getFracTimeMcu();
	frac += _fraction;
	return frac * ((float)1000 / (float) RTC_WAKE_INTERVAL_1S);
}

uint32_t RTC_GetIntervalSeconds(uint32_t seconds)
{
	seconds = (seconds > RTC_SCHEDULE_MAX_SEC) ? RTC_SCHEDULE_MAX_SEC : seconds;
	seconds = (seconds < RTC_SCHEDULE_MIN_SEC) ? RTC_SCHEDULE_MIN_SEC : seconds;
	return seconds;
}

uint32_t RTC_GetTimeDiff(uint32_t time){
	if(time > _seconds)
		return 0;
		
	return _seconds - time;
}

int64_t RTC_GetTimeDiffMs(uint32_t timeSec, uint32_t ms){
	int64_t seconds = (int64_t)RTC_GetCurrentTime();
	int64_t fraction = (int64_t)RTC_GetCurrentFractionalMs();

	return (int64_t)(((seconds - timeSec) * 1000) - (int64_t)ms + (int64_t)fraction);
}

void RTC_ScheduleStart(RtcSchedule_t *schedule, uint32_t seconds, void *callBackFunc){
	if(!_initialized)
		RTC_Init(RTC_WAKE_INTERVAL_1S);
		
	rtc_scheduleAdd(schedule);

	seconds = RTC_GetIntervalSeconds(seconds);
	schedule->Interval = seconds;
	schedule->CallbackFunc = callBackFunc;
	schedule->NextTime = rtc_scheduleGetNextTime(seconds);
	schedule->State = SCHEDULE_RUNNING;
}

//==============================================================================
// Private Functions
//==============================================================================
static void rtc_initMcu(uint8_t enable){			
	if(enable){
		//start the RTC
		NRF_CLOCK->LFCLKSRC = 0x01;	//configured for external XTAL on RTC
		NRF_CLOCK->TASKS_LFCLKSTART = 1;	
		while(!NRF_CLOCK->EVENTS_LFCLKSTARTED);
	
		//Disable RTC
		NRF_RTC0->TASKS_STOP = 1;
	
		//prescale by 8 : 32768 / 8 = 4096 Hz
		NRF_RTC0->PRESCALER = 7;
		
		//define the wake interval
		NRF_RTC0->CC[0] = _interval;
		
		//connect the interrupt event signal on CC[0] compare match
		NRF_RTC0->INTENSET |= 0x00010000;
	
		NVIC_EnableIRQ(RTC0_IRQn);
	
		//Enable RTC
		NRF_RTC0->TASKS_START = 1;
	}
	else{
		NRF_RTC0->TASKS_STOP = 1;
		NRF_RTC0->TASKS_CLEAR = 0;
		_initialized = false;
	}

	//Set current time to 0
	_seconds = 0;
	_fraction = 0;
}

static inline uint16_t rtc_getFracTimeMcu(){
	//the internals to the MCU ensure a safe read
	return NRF_RTC0->COUNTER;
}

static void rtc_scheduleAdd(RtcSchedule_t *schedule){
	if(rtc_scheduleExists(schedule))
		return;
	
	RtcSchedule_t *lastSchedule = _firstSchedule;
	schedule->NextSchedule = NULL;
	
	if(lastSchedule == NULL)
		_firstSchedule = schedule;
	else{
		while(lastSchedule->NextSchedule)
			lastSchedule = lastSchedule->NextSchedule;
		
		lastSchedule->NextSchedule = schedule;
	}
}

static uint8_t rtc_scheduleExists(RtcSchedule_t *schedule){
	RtcSchedule_t *s = _firstSchedule;

	if(s == NULL)
		return false;
	
	for(; s; s = s->NextSchedule){
		if(s == schedule)
			return true;
	}
	
	return false;
}

static uint32_t rtc_scheduleGetNextTime(uint32_t interval){
	if(interval > RTC_SCHEDULE_MAX_SEC)
		interval = RTC_SCHEDULE_MAX_SEC;

	return RTC_GetCurrentTime() + interval;
}

//==============================================================================
// Task Handler
//==============================================================================
void RTC_ScheduleTaskHandler(){
	static uint32_t lastSeconds;

	uint32_t seconds = RTC_GetCurrentTime();
	
	if(lastSeconds == seconds)
		return;
	
	RtcSchedule_t *s = _firstSchedule;
	while (s){
		if(s->State == SCHEDULE_RUNNING && s->NextTime <= seconds)
		{
			scheduleCbFunction func = (scheduleCbFunction)s->CallbackFunc;
			s->NextTime = rtc_scheduleGetNextTime(s->Interval);
			if(func)
				func(s);
		}
		s = s->NextSchedule;
	}
	
	lastSeconds = seconds;
}

//==============================================================================
// Interrupt Handler
//==============================================================================
void RTC0_IRQHandler(void){
	if(NRF_RTC0->EVENTS_COMPARE[0]){
		_fraction += _interval;
		if(_fraction >= RTC_WAKE_INTERVAL_1S)	{
			_seconds += (_fraction / RTC_WAKE_INTERVAL_1S);
			_fraction = 0;
		}
			
		NRF_RTC0->EVENTS_COMPARE[0] = 0;
		NRF_RTC0->TASKS_CLEAR = 1;
	}
}